src/analysis/inner_product_space/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,253 @@
+
+/-- 'gram_schmidt_def' turns the sum over `fin n` into a sum over `ℕ`. -/
+lemma gram_schmidt_def (f : ℕ → E) (n : ℕ) :
+  gram_schmidt 𝕜 f n = f n - ∑ i in finset.range n,
+    orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) :=
+begin
+  rw gram_schmidt,
+  congr' 1,
+  exact fin.sum_univ_eq_sum_range (λ i,
+    (orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) : E)) n,
+end
+
+@[simp] lemma gram_schmidt_zero (f : ℕ → E) :
+  gram_schmidt 𝕜 f 0 = f 0 :=
+by simp only [gram_schmidt, fintype.univ_of_is_empty, finset.sum_empty, sub_zero]
+
+/-- **Gram-Schmidt Orthogonalisation**
+Gram-Schmidt process produces an orthogonal system of vectors. -/
+theorem gram_schmidt_orthogonal (f : ℕ → E) (a b : ℕ) (h₀ : a ≠ b) :
+  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
+begin
+  suffices : ∀ a b : ℕ, a < b → ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0,
+  { cases h₀.lt_or_lt with ha hb,
+    { exact this _ _ ha, },
+    { rw inner_eq_zero_sym,
+      exact this _ _ hb, }, },
+  clear h₀ a b,
+  intros a b h₀,
+  obtain ⟨c, hbc⟩ : ∃ c, b ≤ c := ⟨b, le_rfl⟩,
+  revert a b,
+  apply nat.strong_induction_on c; clear c,
+  intros c hc a b h₀ hbc,
+  rw le_iff_lt_or_eq at hbc,
+  rcases hbc with (hbc | rfl),
+  { exact hc b hbc a b h₀ le_rfl, },
+  simp only [gram_schmidt_def 𝕜 f b, inner_sub_right, inner_sum,
+    orthogonal_projection_singleton, inner_smul_right],
+  rw finset.sum_eq_single_of_mem a (finset.mem_range.mpr h₀),
+  { by_cases h : gram_schmidt 𝕜 f a = 0,
+    { simp only [h, inner_zero_left, zero_div, zero_mul, sub_zero], },
+    { rw [← inner_self_eq_norm_sq_to_K, div_mul_cancel, sub_self],
+      rwa [ne.def, inner_self_eq_zero], }, },
+  intros i hi hia,
+  simp only [mul_eq_zero, div_eq_zero_iff, inner_self_eq_zero],
+  right,
+  rw finset.mem_range at hi,
+  cases hia.lt_or_lt with hia₁ hia₂,
+  { rw inner_eq_zero_sym,
+    exact hc a h₀ i a hia₁ le_rfl, },
+  { exact hc i hi a i hia₂ le_rfl, },
+end
+
+open submodule set
+
+/-- Gram-Schmidt process preserves span -/
+lemma span_gram_schmidt (f : ℕ → E) (c : ℕ) :
+  span 𝕜 (gram_schmidt 𝕜 f '' Iic c) = span 𝕜 (f '' Iic c) :=
+begin
+  induction c with c hc,
+  { simp only [Iic, gram_schmidt_zero, le_zero_iff,
+      set_of_eq_eq_singleton, image_singleton], },
+  have h : Iic c.succ = insert c.succ (Iic c),
+  { ext,
+    simp only [set.Iic, mem_set_of_eq, mem_insert_iff,
+      nat.lt_succ_iff, le_iff_lt_or_eq, or_comm], },
+  have h₀ : ∀ b, b ∈ finset.range c.succ → gram_schmidt 𝕜 f b ∈ span 𝕜 (f '' Iic c),
+  { intros b hb,
+    rw [finset.mem_range, nat.succ_eq_add_one] at hb,
+    have hb₁ : b ≤ c := by linarith,
+    have hb₂ : gram_schmidt 𝕜 f b ∈ span 𝕜 (gram_schmidt 𝕜 f '' Iic c),
+      { have h₀ : gram_schmidt 𝕜 f b ∈ gram_schmidt 𝕜 f '' Iic c,
+        { simp only [mem_image, mem_Iic],
+          refine ⟨b, hb₁, by refl⟩, },
+          have h₁ : gram_schmidt 𝕜 f '' Iic c ≤ span 𝕜 (gram_schmidt 𝕜 f '' Iic c) := subset_span,
```

Comment:
Indent is wrong here, you're not inside the `{}` on the previous line any more

Reply:
oh yes, sorry for not careful here, I have fixed it, thank you!